draft-ietf-quic-transport.md

```diff
@@ -732,26 +764,39 @@ sequence of frames, as described in {{frames}}.
 
 ## Connection ID {#connection-id}
 
-QUIC connections are identified by their 64-bit Connection ID.  All long headers
-contain a Connection ID.  Short headers indicate the presence of a Connection ID
-using the Omit Connection ID flag.  When present, the Connection ID is in the
-same location in all packet headers, making it straightforward for middleboxes,
-such as load balancers, to locate and use it.
-
-The client MUST choose a random connection ID and use it in Initial packets
-({{packet-initial}}) and 0-RTT packets ({{packet-protected}}).
-
-When the server receives a Initial packet and decides to proceed with the
-handshake, it chooses a new value for the connection ID and sends that in a
-Retry ({{packet-retry}}) or Handshake ({{packet-handshake}}) packet.  The server
-MAY choose to use the value that the client initially selects.
-
-Once the client receives the connection ID that the server has chosen, it MUST
-use it for all subsequent Handshake ({{packet-handshake}}) and 1-RTT
-({{packet-protected}}) packets but not for 0-RTT packets ({{packet-protected}}).
-
-Server's Version Negotiation ({{packet-version}}) and Retry ({{packet-retry}})
-packets MUST use connection ID selected by the client.
+A connection ID is used to ensure consistent routing of packets.  The long
+header contains two connection IDs: the Destination Connection ID is chosen by
+the recipient of the packet and is used to provide consistent routing; the
+Source Connection ID is used to set the Destination Connection ID used by a
+peer.
+
+During the handshake, packets with the long header are used to establish the
+Destination Connection ID that each endpoint uses.  Each endpoint uses the
+Source Connection ID field to specify the connection ID that is used in the
+Destination Connection ID field of packets being sent to them.  Upon receiving a
+packet, each endpoint sets the Destination Connection ID it sends to match the
+value of the Source Connection ID that they receive.
+
+During the handshake, an endpoint might receive multiple packets with the long
+header, and thus be given multiple opportunities to update the Destination
+Connection ID it sends.  A client MUST only change the value it sends in the
+Destination Connection ID field in response to the first packet of each type
+(Retry, or Handshake) that it receives; a server MUST only change its value
```

Comment:
I think an algorithm based on packet number might be more robust than specifying the first packet of each type?  Largest packet number wins?

Also, I would say the server MUST set it's value based on an Initial packet.  Given there was no connection before, I wouldn't call it a change of connection ID.

Reply:
+1. I can't reason about packet type changes easily, but largest packet number seems robust.

Reply:
There is a discussion on list about version negotiation attacks by man-on-the-side. One proposed solution was to queue some packets and wait for the best outcome. I'm not so sure what is best, but it overlaps with the considerations in the above text and comments.

Reply:
Largest packet number fails for the simple reason that it might allow for multiple changes.  Lowest packet number works better.  And that's all I'm looking for.  Latching is what we are looking for, not a continuous sequence of changes.

Also, I added a note to ignore packets with different Source Connection ID values.  That deals neatly with the problem where two server instances both attempt to start a connection with you as a result of you sending multiple Initial packets.